R/audits.R

```diff
@@ -1,110 +1,49 @@
-#' @export
-cache_apps <- function(connect) {
-  apps <- connect$get_apps()
-  apps
-}
-
-#' Audit Vanity URLs
+#' Check to see if a vanity URL is currently in use
 #'
 #' \lifecycle{experimental}
 #'
-#' @param apps App list, see [cache_apps()]
-#' @param server_url Base url for the Connect server
-#' @param vanity Optional, see details
+#' @param connect A Connect R6 object
+#' @param vanity string of the vanity URL to check
 #'
-#' @details If `vanity` is not provided, returns a list of all the vanity
-#'   urls in use. If `vanity` is provided, returns whether or not
-#'   `vanity` is eligible as a vanity url.
+#' @return logical indicating if the vanity URL is available.
 #'
 #' @family audit functions
 #' @export
-audit_vanity_urls <- function(apps, server_url, vanity = NULL) {
-  # TODO: why does vanities not work?
-  urls <- get_field(apps, "url")
-  parse_server <- httr::parse_url(server_url)
-  if (is.null(parse_server$scheme)) {
-    stop(glue::glue("ERROR: protocol (i.e. http:// or https://) not defined on server_url={server_url}"))
-  }
-  content <- sapply(urls, function(u) {
-    trim_vanity(u, parse_server$path)
-  })
+vanity_is_available <- function(connect, vanity) {
+  current_vanities <- connect$GET(v1_url("vanities"))
+  current_vanity_paths <- purrr::map_chr(current_vanities, "path")
 
-  vanities <- content[!grepl("content/\\d+", content)]
+  # In case a full URL has been given, prune it down to just the path
+  vanity <- sub(paste0("^", connect$server), "", vanity)
+  # and make sure it has a leading and trailing slash
+  vanity <- sub("^/?(.*[^/])/?$", "/\\1/", vanity)
 
-  if (!is.null(vanity)) {
-    return(
-      ifelse(sprintf("%s/", vanity) %in% vanities, sprintf("%s Not Available", vanity), sprintf("%s Available", vanity))
-    )
-  }
-  vanities
+  !(vanity %in% current_vanity_paths)
 }
 
-trim_vanity <- function(url, server_path) {
-  parsed_url <- httr::parse_url(url)
-  if (nchar(server_path) > 0) {
-    # remove the trailing slash
-    server_path <- base::sub("^(.*)/$", "\\1", server_path)
-    vanity <- sub(server_path, "", parsed_url$path)
-  } else {
-    vanity <- parsed_url$path
-  }
-
-  # ensure leading and trailing slash
-  base::sub("^/?(.*[^/])/?$", "/\\1/", vanity)
-}
-
-
 #' Audit R Versions
 #'
 #' \lifecycle{experimental}
 #'
-#' @param apps App list, see `cache_apps`
+#' @param content `data.frame` of content information, as from [get_content()]
 #'
 #' @return A plot that shows the R version used by content over time and in
 #'   aggregate.
 #' @family audit functions
 #' @export
-audit_r_versions <- function(apps) {
-  r_versions <- get_field(apps, "r_version", TRUE)
-  published <- get_field(apps, "last_deployed_time", TRUE)
+audit_r_versions <- function(content) {
+  if (!requireNamespace("ggplot2", quietly = TRUE)) {
+    stop("ggplot2 is required for this function")
+  }
+  if (!requireNamespace("gridExtra", quietly = TRUE)) {
+    stop("gridExtra is required for this function")
+  }
```

Comment:
I wonder if we should make this a helper since we are hoping to prune things down more like this. Maybe not here, but seeing them back to back and knowing other places we might use it...

Reply:
There's probably an rlang helper for this.

Reply:
https://rlang.r-lib.org/reference/is_installed.html I guess we could do `rlang::check_installed(c("ggplot2", "gridExtra"))`? The interactive behavior of asking you if you want to install them seems odd to me but maybe that's just me.